### ISSUE TITLE
[css-contain-4] Define a range syntax for style container queries

### ISSUE THREAD

**Issue by mirisuzanne** [w3c/csswg-drafts#8376]
The css-contain-3 spec provides a '[style container feature](https://www.w3.org/TR/css-contain-3/#style-container)', allowing authors to query the computed value of a given property on the container. However, that query is currently limited to the declaration syntax, which provides a simple equality comparison:

```css
@container style(--color-mode: dark) { … }
```

As discussed in #7068, there are also use-cases for range comparisons on container properties:

```css
@container style(padding > 1em) { ... }
@container style(--primary-color-lightness > 80) { ... }
```

For custom properties, this might require a defined syntax using `@property`? Or would we want a way to define the value type on the query itself? Thoughts?

**Comment by LeaVerou** [comment]
For syntax, we can just adopt the MQ operators.

I think it makes sense to require these properties to be registered as `<integer>`, `<number>`, `<length>`, `<percentage>`, or `<length-percentage>` (am I forgetting anything?), we can always expand later.

Or, crazy idea here, given this is fairly limited and these types can be detected syntactically, what if we don't require registration at all? Would that be possible? Is there any ambiguity I'm missing?

**Comment by SebastianZ** [comment]
> I think it makes sense to require these properties to be registered as `<integer>`, `<number>`, `<length>`, `<percentage>`, or `<length-percentage>` (am I forgetting anything?)

`<angle>`, `<frequency>`, `<time>` and their `<*-percentage>` counter-parts, `<ratio>` and maybe some more.

> For syntax, we can just adopt the MQ operators.

I assume that includes the full range syntax, so one could write

```css
@container style(1em < padding < 3em) { … }
```

Correct?

Sebastian

**Comment by tabatkins** [comment]
There is indeed no ambiguity when limited to comparisons of this type - there's no keywords or similar that can be interpreted differently on context, just plain numbers and dimensions.

And math functions, which is the only thing that makes me slightly concerned, just due to the potential complexity. But I suspect it's fine.

**Comment by mirisuzanne** [comment]
Another complexity: is it allowed to mention multiple properties? `style(margin < 3em < padding) { … }`

**Comment by LeaVerou** [comment]
> Another complexity: is it allowed to mention multiple properties? `style(margin < 3em < padding) { … }`

I'd suggest not, at first.

**Comment by SebastianZ** [comment]
> Another complexity: is it allowed to mention multiple properties? `style(margin < 3em < padding) { … }`

That syntax doesn't make sense to me. I'd say we should stick to the syntax defined for media queries. So that could be expressed as `style(margin < 3em) and style(padding > 3em)`. This also makes it more readable, in my opinion.

Sebastian

**Comment by tabatkins** [comment]
Yeah, agreed that we should stick to the MQ syntax for now - either a property compared to a literal, or a property sandwiched between two literal comparisons.

**Comment by mirisuzanne** [comment]
I marked this for contain-4, but not sure if it needs to be deferred from l3? It doesn't seem like a big lift on the spec side, but may be a larger consideration for implementors?

**Comment by una** [comment]
I just mentioned this in the older [thread](https://github.com/w3c/csswg-drafts/issues/7068#issuecomment-1414426214), but this would be a very useful feature. And +1 to the MQ range syntax as a mechanism to query values by.

**Comment by una** [comment]
We would need to limit (and define the limitations) of the syntax options for range queries? I.e.

**Valid:** 
- `<length>`
- `<number>`
- `<percentage>`
- `<length-percentage>`
- `<color>`
- `<integer>`
- `<angle>`
- `<time>`
- `<resolution>`

**Invalid:**
- `<image>`
- `<url>`
- `<custom-ident>`
- `<transform-function>`
- `<transform-list>`

**Comment by tabatkins** [comment]
Yes, it would be all the [numeric types](https://drafts.csswg.org/css-values/#numeric-data-types). Not colors, tho.

**Comment by mirisuzanne** [comment]
@lilles raised a question this morning about how to resolve `<length-percentage>` on custom properties - where it's not clear how percentages compare to numbers.

**Comment by tabatkins** [comment]
Percents aren't comparable to lengths in contexts where they don't resolve against a length, so they'd just be false, I suppose?  That is, `--foo > 50%` can be true if `--foo: 75%`, but if `--foo: 75px` it'll always be false. This should be the same behavior as comparing a `<length>` custom property to an angle; we can't tell at parse time whether the comparison is syntactically valid, so all we can do is say it's false when we actually make the comparison.

(Do we want to try and do syntax validation on known properties? Or should we just consider them the same way, where doing a nonsensical comparison like `padding-left > 50deg` is false?)

**Comment by mirisuzanne** [comment]
I think it would be ok to do without parse-time syntax validation - but I would probably expect `unknown` rather than `false` in both cases.

**Comment by tabatkins** [comment]
We don't use unknown for "this doesn't make sense", we use it for "this conceivably has a true/false answer, but we happen to not know what it is; a future browser might, tho". So nonsense comparisons are simply false, not unknown. 

And in a custom property, %s *do not* resolve against anything, so they're incomparable with any other types, just as much as lengths and angles are incomparable.

**Comment by andruud** [comment]
Are we going with the idea that registration isn't required, but the custom property side is interpreted according to the type of the literal side?

**Comment by tabatkins** [comment]
Right; registration doesn't add anything to the equation in this case. If we see `--foo > 2em`, we know that we're gonna be comparing the property value to a length, so we can interpret it that way immediately.

I mean, maybe registration lets you fast-path a false for nonsensical comparisons, but that's it.

Tho hm, do we ever want to allow something like `--foo > --bar`? In that case, registration would be useful.

**Comment by andruud** [comment]
Great, just wanted to clarify that so we know what to implement.

> Tho hm, do we ever want to allow something like --foo > --bar? In that case, registration would be useful.

It could work without as well? We just try to parse a numeric thing on both sides, and check if they're compatible.

**Comment by tabatkins** [comment]
> We just try to parse a numeric thing on both sides, and check if they're compatible.

Yeah true.

**Comment by mirisuzanne** [comment]
Does this logic seem like something we could add to css-contain-3, or would need to be deferred to L4?

**Comment by tabatkins** [comment]
This also seems to have implications for the non-range syntax, yeah? Like, we'd want `style(--x: 1.0)` to match an `--x: 1;`, because the two are numerically equal. But `style(--x: 1.0 foo)` wouldn't match `--x: 1 foo;`, because the value isn't a single numerical value, so we'll instead use serialization matching.

(I think this is fine, fwiw.)

**Comment by andruud** [comment]
Should we consider making `:` _always_ serialization-matching, and `=` _only_ "recognized-type"-matching?

**Comment by mirisuzanne** [comment]
I like the consistency, and since authors have both options, they can opt into either behavior.

**Comment by andruud** [comment]
> Right; registration doesn't add anything to the equation in this case.

...

> > We just try to parse a numeric thing on both sides, and check if they're compatible.

...

> consider making : always serialization-matching, and = only "recognized-type"-matching?

---

This has marinated for ~two years: can we make progress here now?

**Proposal:**

 - We adopt MQ's syntax, including the sandwich form (e.g. a < b < c).
 - Every part can be a literal, or a reference to a custom property.
 - `=` matches by value (a number `1.0` is equal to a number `1`), whereas `:` matches by serialization (a value `1.0` is _not_ equal to a value `1`).

For e.g. style(A > B):

 - If A/B is a custom property reference, substitute with the value of the custom property.
 - Parse each side according to `<number> | <percentage> | <length> | <angle> | <time> | <frequency> | <resolution>`.
 - If both sides have the same type, compute each side, do the comparison, and return the result.
 - Otherwise, return false.

Note that this allows `style(--foo < --bar)`, but also `style(10px < 10em)` which we _could_ disallow if we think it doesn't fit here.

Note also that you can't compare a `<length>` to a unitless zero in in this approach, e.g. `style(--length > 0)` does not work: it needs to be `style(--length > 0px)`. We could special-case this if we wanted, by considering zero implicitly compatible with other numeric types for comparison purposes.

**Comment by kizu** [comment]
> but also `style(10px < 10em)` which we could disallow if we think it doesn't fit here.

I don't think there is any reason to disallow this, as this can be a quite common use. Probably not on its own (although, you _could_ want to check the value of `1em`), but when used with mixed custom properties.

> Note also that you can't compare a `<length>` to a unitless zero in in this approach, e.g. `style(--length > 0)` does not work: it needs to be `style(--length > 0px)`. We could special-case this if we wanted, by considering zero implicitly compatible with other numeric types for comparison purposes.

Hmm, as an author, I would expect the `0 = 0px` to be true, in the same way `1.0` will be equal to `1`, _as long as we know the type_. Like when we have a style query with a registered `<length>` custom property, a `0` and `0px` comparison will be always true: https://codepen.io/kizu/pen/ByaPWQv, so I'd vote for special-casing `0` to be more flexible.

**Comment by mirisuzanne** [comment]
I like the proposal, and would love for it to move forward.

> `style(10px < 10em)`

There would be other versions of this that are less useful (px vs inch?) - but this seems like it would have solid use cases, since `em` are resolved against the container font size.

> I'd vote for special-casing 0 to be more flexible.

Me too

**Comment by andruud** [comment]
> Hmm, as an author, I would expect the `0 = 0px` to be true, in the same way `1.0` will be equal to `1`, _as long as we know the type_. Like when we have a style query with a registered `<length>` custom property, a `0` and `0px` comparison will be always true: https://codepen.io/kizu/pen/ByaPWQv

Sure, OK. Note, however, that this is more like `<number> | <length>`:

```css
@property --x {
  syntax: "<number> | <length>";
  initial-value: 0px;
  inherits: false;
}
#child {
  color: red;
  @container style(--x: 0) { /* false */
    color: green;
  } 
}
```

But we can certainly still make `0 = 0px` work.

**Comment by andruud** [comment]
> > but also `style(10px < 10em)` which we could disallow if we think it doesn't fit here.
> 
> I don't think there is any reason to disallow this, as this can be a quite common use. Probably not on its own (although, you _could_ want to check the value of `1em`), but when used with mixed custom properties.

Then I guess we'd also allow e.g. `style(var(--x) < var(--y))`? If yes, then e.g. `style(--x < 10px)` is basically syntax sugar for `style(var(--x) < 10px)`.

**Comment by mirisuzanne** [comment]
That seems helpful because it also allows variable fallbacks in the comparisons.

**Comment by andruud** [comment]
> > but also `style(10px < 10em)` which we could disallow if we think it doesn't fit here.
> 
> I don't think there is any reason to disallow this, as this can be a quite common use. Probably not on its own (although, you _could_ want to check the value of `1em`), but when used with mixed custom properties.

Given https://github.com/w3c/csswg-drafts/issues/11104, is it dumb that both `if((10px < 1em):value)` and `if(style(10px < 1em):value)` are allowed?

**Comment by mirisuzanne** [comment]
> Given [#11104](https://github.com/w3c/csswg-drafts/issues/11104), is it dumb that both `if((10px < 1em):value)` and `if(style(10px < 1em):value)` are allowed?

@andruud Are both resolving relative lengths on the same element? I don't know if it's a good or bad idea, but I'm curious if they mean the same thing?

**Comment by andruud** [comment]
Yes, as far as I understand, they mean the same thing.

**Comment by mirisuzanne** [comment]
For discussion sake, the proposal is here: https://github.com/w3c/csswg-drafts/issues/8376#issuecomment-2751161553 - follow-up comments provide some additional details and clarifications.

**Comment by tabatkins** [comment]
If we're accepting `style(10px < 1em)`, then we can edit the resolution in #11104 to just use `style()` for that use-case.
